### PR TITLE
adds direct transcription for discrete-time systems

### PR DESCRIPTION
**Pull request by RussTedrake**
with unit test coverage.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6887)

### CODE REVIEW

**Comment by RussTedrake** [comment]
+@avalenzu for feature review, please.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-:-KsCg4itm8RoHfEIIqMp:b-fcohnv)*

**Comment by avalenzu** [comment]
Review pass done. PTAL.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/trajectory_optimization/BUILD, line 69 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJeGc5AgmEwUKKrLEo:-KsJeGc5AgmEwUKKrLEp:b-lvy6fh) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/BUILD#L69)):*
> ```Python
>     deps = [
>         ":direct_collocation",
>         "//drake/common:eigen_matrix_compare",
> ```

Do you actually want to change the direct collocation target in this PR?

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJTdbPCuQbBQnQlYBc:-KsJTdbPCuQbBQnQlYBd:b-rurnbz) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L47)):*
> 2 * num_states_ + num_inputs_

This could be replaced with `num_vars()`

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJU4nGDZ5vx8zAB0cT:-KsJU4nGDZ5vx8zAB0cU:b-b6as7u) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L66)):*
> ```cpp
> 
>     // Extract our input variables:
>     const auto input = x.segment(0, num_inputs_);
> ```

Consider replacing with `head(num_inputs)` to match use of `tail()` below

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 99 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJV0ZOTMRM7CCmYT5o:-KsJV0ZOTMRM7CCmYT5p:b-7eckc6) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L99)):*
> ```cpp
>     : MultipleShooting(system->get_num_total_inputs(),
>                        context.get_num_total_states(), num_time_samples,
>                        0.1),  // TODO(russt): Replace this with the actual
> ```

Reference #6878 here.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJYAjlgD-w8imk75yO:-KsJYAjlgD-w8imk75yP:b-7eckc6) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L108)):*
> ```cpp
>   DRAKE_THROW_UNLESS(context.has_only_discrete_state());
> 
>   // TODO(russt): Check that the system has ONLY simple periodic updates.
> ```

Reference #6878 here.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsED1U4JdYBml5mZdOA:-KsED1U4JdYBml5mZdOB:b4ijba8) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L17)):*
> where we have decision variables representing every timestep of the trajectory, and one-step of numerical integration provides the dynamic constraints between those decision variables.

I'd suggest rewording this as (modulo "timestep" vs. "knot point" question from above)

> where decision variables represent the state and input at every timestep of the trajectory, and one-step of numerical integration provides the dynamic constraints between adjacent timesteps.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJOYlhpm6GKecNgrHu:-KsJOYlhpm6GKecNgrHv:b-wraajc) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L17)):*
> ```objc
> /// DirectTranscription is perhaps the simplest implementation of a multiple
> /// shooting method, where we have decision variables representing every
> /// timestep of the trajectory, and one-step of numerical integration
> ```

FYI: I've created #6901 to cover standardizing our naming for "the points at which the trajectories are discretized"

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsECqMO9YQrzY5tJqxj:-KsECqMO9YQrzY5tJqxk:b1yma0d) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L35)):*
> ```objc
>   ///    values of the state in this context do not have any effect.  This
>   ///    context will also be "cloned" by the optimization; changes to the
>   ///    context
> ```

Stray line break here.

---

*[drake/systems/trajectory_optimization/test/direct_collocation_test.cc, line 80 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXAsdpWNqnlwsb5Pb:-KsJXAsdpWNqnlwsb5Pc:bhq7xqo) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_collocation_test.cc#L80)):*
> ```cpp
>   const auto context = system->CreateDefaultContext();
> 
>   const int kNumSampleTimes = 4;
> ```

Are the changes to this file actually related to this PR?

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXWaTdfa_Y4LIWF7p:-KsJXWaTdfa_Y4LIWF7q:boqurmm) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L22)):*
> DirectTimeConstructorThrows

Should this be `DiscreteTimeConstructorThrows`?

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXdvwF_lOnod423sU:-KsJXdvwF_lOnod423sV:b4fv7jw) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L50)):*
> ```cpp
>   DirectTranscription prog(&system, *context, kNumSampleTimes);
> 
>   // TODO(russt):  Uncomment this pending resolution of #6878.
> ```

s/pending/upon/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-:-KsEDg0kACYY4mehXwEf:b837gv3)*

**Comment by RussTedrake** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/trajectory_optimization/BUILD, line 69 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJeGc5AgmEwUKKrLEo:-KsLiA9KNXyQqJs7GySc:beuhkbo) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/BUILD#L69)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Do you actually want to change the direct collocation target in this PR?
</blockquote></details>

there is no change.  just an ugly printout from diff.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJTdbPCuQbBQnQlYBc:-KsLiNvRpTk2FyUYkLHP:b5u615l) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L47)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> 2 * num_states_ + num_inputs_

This could be replaced with `num_vars()`
</blockquote></details>

that seems too fragile.  what if i came around and added some other decision variable (admittedly, would have to be in the constructor) before constructing the dynamic constraints?

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 66 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJU4nGDZ5vx8zAB0cT:-KsLikTFL_5upadV8hnK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L66)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Consider replacing with `head(num_inputs)` to match use of `tail()` below
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 99 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJV0ZOTMRM7CCmYT5o:-KsLj2Q6Je7j0LX2Lxgj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L99)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Reference #6878 here.
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 108 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJYAjlgD-w8imk75yO:-KsLj8GQOaeMGpvAwfBO:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L108)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Reference #6878 here.
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsED1U4JdYBml5mZdOA:-KsLj_vU1tFmSLb7xB7y:bbbinsh) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L17)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> where we have decision variables representing every timestep of the trajectory, and one-step of numerical integration provides the dynamic constraints between those decision variables.

I'd suggest rewording this as (modulo "timestep" vs. "knot point" question from above)

> where decision variables represent the state and input at every timestep of the trajectory, and one-step of numerical integration provides the dynamic constraints between adjacent timesteps.

</blockquote></details>

happy to continue cleaning.  but it was broken before and I've made it better.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJOYlhpm6GKecNgrHu:-KsLjd_Q2Wrx0VP3B913:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L17)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

FYI: I've created #6901 to cover standardizing our naming for "the points at which the trajectories are discretized"
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 35 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsECqMO9YQrzY5tJqxj:-KsLjlkTxwcne0Azp_xC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L35)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Stray line break here.
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/test/direct_collocation_test.cc, line 80 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXAsdpWNqnlwsb5Pb:-KsLjtT4P8aCwxvahL7q:bxhi5hj) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_collocation_test.cc#L80)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Are the changes to this file actually related to this PR?
</blockquote></details>

i admit i'm (intentionally) sneaking them in.  my testing revealed that I needed the tolerance.  

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 22 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXWaTdfa_Y4LIWF7p:-KsLkIWz13V4j05POBHP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L22)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

> DirectTimeConstructorThrows

Should this be `DiscreteTimeConstructorThrows`?
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/test/direct_transcription_test.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXdvwF_lOnod423sU:-KsLkMMMiEBU7JTpN9cg:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_transcription_test.cc#L50)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

s/pending/upon/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887)*

**Comment by RussTedrake** [comment]
+@soonho-tri for platform review, please.  (still want @avalenzu's approval, but would also like to catch soonho before his window closes)

---

Review status: 4 of 7 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-:-KsLlINzT8cNfY2j6Gmy:b-h84z4w)*

**Comment by RussTedrake** [comment]
Review status: 3 of 7 files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/trajectory_optimization/test/direct_collocation_test.cc, line 80 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXAsdpWNqnlwsb5Pb:-KsLmedX-zc_z43aIClT:b-d7gfth) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_collocation_test.cc#L80)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i admit i'm (intentionally) sneaking them in.  my testing revealed that I needed the tolerance.  
</blockquote></details>

fwiw - i've separated them out to a separate commit.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887)*

**Comment by avalenzu** [comment]
:lgtm: Excited to see this go in!

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/trajectory_optimization/BUILD, line 69 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJeGc5AgmEwUKKrLEo:-KsOZcg38iWAAfikQGLt:b-cphear) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/BUILD#L69)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

there is no change.  just an ugly printout from diff.
</blockquote></details>

You're completely right. Sorry about that.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 47 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJTdbPCuQbBQnQlYBc:-KsOZomfN-zf3xFVwYXl:b-xhapn8) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.cc#L47)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

that seems too fragile.  what if i came around and added some other decision variable (admittedly, would have to be in the constructor) before constructing the dynamic constraints?
</blockquote></details>

If you added a decision variable that affected the dynamics (e.g. corresponding to a parameter of the system) wouldn't you want that to be accounted for here? And if you added decision variables that didn't affect the dynamics, would it be wrong to carry around zero gradients for them? Or just a performance hit?

It seems like an opportunity for future error to compute `2 * num_states_ + num_inputs_` both in the call to the `Constraint` constructor and here. That being said, I'm fine with leaving this as is if you think that's best.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 17 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsED1U4JdYBml5mZdOA:-KsO_vDKxC1DTkBrbSVt:byn2t1q) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/direct_transcription.h#L17)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

happy to continue cleaning.  but it was broken before and I've made it better.
</blockquote></details>

This is much better now.

---

*[drake/systems/trajectory_optimization/test/direct_collocation_test.cc, line 80 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsJXAsdpWNqnlwsb5Pb:-KsOa8BrLbfx3XToQ_wK:beehhyy) ([raw file](https://github.com/robotlocomotion/drake/blob/799925418b19d635dce9654c7d96ba0b060df5d4/drake/systems/trajectory_optimization/test/direct_collocation_test.cc#L80)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

fwiw - i've separated them out to a separate commit.  
</blockquote></details>

That's fine with me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-:-KsOb_NRQ4h9LShRzWMr:b-vtflmf)*

**Comment by soonho-tri** [comment]
:lgtm:

---

Reviewed 3 of 7 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP8GF-8Ni5rvP-QNlT:-KsP8GF-8Ni5rvP-QNlU:b8hikf8) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.cc#L23)):*
> ```cpp
>   // The format of the input to the eval() function is a vector
>   // containing {input, state, next_state}.
> ```

FYI, these two lines can be removed.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP9y7CYdLWbwXsuqtS:-KsP9y7CYdLWbwXsuqtT:b-eotuq7) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.cc#L51)):*
> ```cpp
>   }
> 
>   virtual ~DiscreteTimeSystemConstraint() {}
> ```

Its destructor is already `virtual` because of `EvaluatorBase` (the parent of `Constraint`). To following our convention, let's have:
```c++
  ~DiscreteTimeSystemConstraint() override = default;
```

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 83 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsPD8QGVCL2vHyJAmyX:-KsPD8QGVCL2vHyJAmyY:b-ei33ed) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.cc#L83)):*
> ```cpp
>  private:
>   const System<AutoDiffXd>& system_;
>   Context<AutoDiffXd>* context_;
> ```

BTW, the following three members can be `const`:
```c++
  Context<AutoDiffXd>* const context_;
  FreestandingInputPortValue* const input_port_value_;
  DiscreteValues<AutoDiffXd>* const discrete_state_;
```

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP7lDY8AStipqgwO43:-KsP7lDY8AStipqgwO44:b2bt0me) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L6)):*
> ```objc
> 
> #include "drake/common/drake_copyable.h"
> #include "drake/solvers/constraint.h"
> ```

FYI, this line can be moved to `direct_transcription.cc` file.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP6LRbTIQAzb-Xl_L1:-KsP6LRbTIQAzb-Xl_L2:bfsrg9l) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L24)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(DirectTranscription)
> 
>   /// Constructs the %MathematicalProgram% and adds the dynamic constraints.
> ```

FYI, as @sherm1 pointed out in another PR, `%MathematicalProgram` is sufficient to suppress auto-linking. Ditto for `%PiecewisePolynomialTrajectory%` in this file.

Reference: http://www.stack.nl/~dimitri/doxygen/manual/faq.html#faq_class

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP5pTbPasDO_DZ3Pw-:-KsP5pTbPasDO_DZ3Pw0:b-h41x4w) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L36)):*
> ```objc
>   ///    values of the state in this context do not have any effect.  This
>   ///    context will also be "cloned" by the optimization; changes to the
>   ///    context after calling this method will NOT impact the trajector
> ```

BTW, nit, s/`trajector`/`trajectory`/

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 63 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP79ZpAw_kdTOABcP6:-KsP79ZqkAAuwqLPTnR9:b-dxu4ye) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L63)):*
> ```objc
> 
>   // Create AutoDiff versions of the System components (for the constraints).
>   const std::unique_ptr<const System<AutoDiffXd>> system_{nullptr};
> ```

FYI, nit, no need to have `{nullptr}` for `std::unique_ptr`s.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-:-KsPXhdR6ifGSkJn6y0a:bnfp4nl)*

**Comment by RussTedrake** [comment]
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP8GF-8Ni5rvP-QNlT:-KsQQDLJbdWNVPC3yuro:b89vajb) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.cc#L23)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, these two lines can be removed.
</blockquote></details>

i think it's valuable, but i moved them down to a more appropriate spot.   fixed in direct_collocation.cc also

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP9y7CYdLWbwXsuqtS:-KsQQTZmzsrfE3EjIteL:b-y67gxp) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.cc#L51)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Its destructor is already `virtual` because of `EvaluatorBase` (the parent of `Constraint`). To following our convention, let's have:
```c++
  ~DiscreteTimeSystemConstraint() override = default;
```
</blockquote></details>

fixed here (and in direct_collocation.cc)

---

*[drake/systems/trajectory_optimization/direct_transcription.cc, line 83 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsPD8QGVCL2vHyJAmyX:-KsQRjdEeQtjRAhTkj5P:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.cc#L83)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, the following three members can be `const`:
```c++
  Context<AutoDiffXd>* const context_;
  FreestandingInputPortValue* const input_port_value_;
  DiscreteValues<AutoDiffXd>* const discrete_state_;
```
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 6 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP7lDY8AStipqgwO43:-KsQRMK7hYsVuHBRXr00:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L6)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, this line can be moved to `direct_transcription.cc` file.
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 24 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP6LRbTIQAzb-Xl_L1:-KsQRFFbAJjMvPp-sQ8e:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L24)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, as @sherm1 pointed out in another PR, `%MathematicalProgram` is sufficient to suppress auto-linking. Ditto for `%PiecewisePolynomialTrajectory%` in this file.

Reference: http://www.stack.nl/~dimitri/doxygen/manual/faq.html#faq_class
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP5pTbPasDO_DZ3Pw-:-KsQR5h6v9D-RlmcBIpb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L36)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, nit, s/`trajector`/`trajectory`/
</blockquote></details>

Done.

---

*[drake/systems/trajectory_optimization/direct_transcription.h, line 63 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6887#-KsP79ZpAw_kdTOABcP6:-KsQR2I-uObgNYj1RVF9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/9da1e789d62d04497a062f043e6e7d872d3e21e6/drake/systems/trajectory_optimization/direct_transcription.h#L63)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, nit, no need to have `{nullptr}` for `std::unique_ptr`s.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6887)*
